src/machine/machine_mimxrt1062.go

```diff
@@ -301,6 +330,101 @@ func (p Pin) Toggle() {
+// interrupts generated on the high-speed GPIO pins.
+//
+// Unfortunately, all four high-speed GPIO ports (A-D) are connected to just a
+// single interrupt control line. Therefore, the interrupt status register (ISR)
+// must be checked in all four GPIO ports on every interrupt.
+func (jt *pinJumpTable) dispatchInterrupt(interrupt.Interrupt) {
+	handle := func(gpio *nxp.GPIO_Type, port Pin) {
+		if status := gpio.ISR.Get() & gpio.IMR.Get(); status != 0 {
+			gpio.ISR.Set(status) // clear interrupt
+			for status != 0 {
+				p := Pin(bits.TrailingZeros32(status))
+				i := Pin(port + p)
+				jt.lut[i](i)
+				status &^= 1 << p
+			}
+		}
+	}
+	if jt.numDefined > 0 {
+		handle(nxp.GPIO6, portA)
+		handle(nxp.GPIO7, portB)
+		handle(nxp.GPIO8, portC)
+		handle(nxp.GPIO9, portD)
+	}
+}
+
+// set associates a function with a given Pin in the receiver lookup table. If
+// the function is nil, the given Pin's associated function is removed.
+func (jt *pinJumpTable) set(pin Pin, fn func(Pin)) {
+	if int(pin) < len(jt.lut) {
+		if nil != fn {
+			if nil == jt.lut[pin] {
+				jt.numDefined++
+			}
+			jt.lut[pin] = fn
+		} else {
+			if nil != jt.lut[pin] {
+				jt.numDefined--
+			}
+			jt.lut[pin] = nil
+		}
+	}
+}
+
+// SetInterrupt sets an interrupt to be executed when a particular pin changes
+// state. The pin should already be configured as an input, including a pull up
+// or down if no external pull is provided.
+//
+// This call will replace a previously set callback on this pin. You can pass a
+// nil func to unset the pin change interrupt. If you do so, the change
+// parameter is ignored and can be set to any value (such as 0).
+func (p Pin) SetInterrupt(change PinChange, callback func(Pin)) error {
+	_, gpio := p.getGPIO() // use fast GPIO for all pins
+	mask := p.getMask()
+	if nil != callback {
+		switch change {
+		case PinLow, PinHigh, PinRising, PinFalling:
+			gpio.EDGE_SEL.ClearBits(mask)
+			var reg *volatile.Register32
+			var pos uint8
+			if pos = p.getPos(); pos < 16 {
+				reg = &gpio.ICR1 // ICR1 = pins 0-15
+			} else {
+				reg = &gpio.ICR2 // ICR2 = pins 16-31
+				pos -= 16
+			}
+			reg.ReplaceBits(uint32(change), 0x3, pos*2)
+		case PinToggle:
+			gpio.EDGE_SEL.SetBits(mask)
+		}
+		pinISR.set(p, callback) // associate the callback with the pin
+		gpio.ISR.Set(mask)      // clear any pending interrupt (W1C)
+		gpio.IMR.SetBits(mask)  // enable external interrupt
+	} else {
+		pinISR.set(p, nil)       // remove any associated callback from the pin
+		gpio.ISR.Set(mask)       // clear any pending interrupt (W1C)
+		gpio.IMR.ClearBits(mask) // disable external interrupt
+	}
+	// enable or disable the interrupt based on number of defined callbacks
+	if pinISR.numDefined > 0 {
+		if nil == pinInterrupt {
+			// create the Interrupt if it is not yet defined
+			irq := interrupt.New(nxp.IRQ_GPIO6_7_8_9, pinISR.dispatchInterrupt)
+			pinInterrupt = &irq
+			pinInterrupt.Enable()
```

Comment:
I don't really understand this, why do you use `&irq`?
```suggestion
			irq.Enable()
```

Reply:
I keep a copy of the interrupt (as _global_ `pinInterrupt`) so that:

1. We don't have to create/enable it when it already exists and is already enabled (`nil != pinInterrupt`)
2. We can disable it when all user callbacks are removed (`pinISR.numDefined == 0`)

I have `pinInterrupt` defined as a pointer (`*interrupt.Interrupt`), and hence use `&irq`, so that I can check if it has been created or not via comparison to `nil`. Wasn't sure of a simple, equivalent way to test a _**value**_ of type `interrupt.Interrupt`.

#### _Related_:

For some reason, `interrupt.New` doesn't follow Go convention and returns a value type instead of a pointer. In this respect, I think it should have been named `interrupt.Make`. In particular, should have been:

> `func New(id int, handler func(Interrupt)) *Interrupt`
> `func Make(id int, handler func(Interrupt)) Interrupt`

Reply:
I also wasn't sure of a way to assign directly to the _pointer_ `pinInterrupt` from the _value_ returned by `interrupt.New`, without using an intermediate variable `irq` and taking its address `&irq`.

Using `pinInterrupt = &interrupt.New(...)`  or `pinInterrupt = &(interrupt.New(...))` isn't valid

Reply:
I think we should probably revisit this specific point, but in a future PR, since there are enough interrelated PRs already! :smiley_cat:

Reply:
Ah I see. Maybe `interrupt.New` is badly named then. It's actually a compiler intrinsic instead of a regular function call, so it will be evaluated at compile time instead of runtime and the return value is compiled down to a constant integer. This magic makes sure that the interrupt handler can be optimized away if the interrupt is never enabled.

Maybe something like `interrupt.Request` is better? Suggestions are definitely welcome.